### PR TITLE
Add an option to trigger flush when the number of range deletions reach a threshold

### PR DESCRIPTION
**Pull request by vrdhn**
Add a mutable column family option `memtable_max_range_deletions`. When non-zero, RocksDB will try to flush the current memtable after it has at least `memtable_max_range_deletions` range deletions. Java API is added and crash test is updated accordingly to randomly enable this option.

Test plan:
* New unit test: `DBRangeDelTest.MemtableMaxRangeDeletions`
* Ran crash test `python3 ./tools/db_crashtest.py whitebox --simple --memtable_max_range_deletions=20` and saw logs showing flushed memtables usually with 20 range deletions.

### CODE REVIEW

File: include/rocksdb/options.h
Comment (cbi42) [inline include/rocksdb/options.h:337]: Can we use range_tombstone/range_deletion to disambiguate from point tombstones? Also maybe include memtable in the name like memtable_max_range_deletion.

File: db/memtable.cc
Comment (yao-xiao-github) [inline db/memtable.cc:793]: Could we set `max_tombstones_reached_ ` here? It would allow the memtable to be flushed without a following read.

File: options/cf_options.h
Comment (yao-xiao-github) [inline options/cf_options.h:318]: I tried patching this PR and run rocksdb. In the logs, I didn't see this option being printed out with other options. Is there anything missing?

File: db/memtable.cc
Comment (yao-xiao-github) [inline db/memtable.cc:174]: It is possible to log the reason for flush? Maybe consider adding the reason to `FlushReason `.

**Review by yao-xiao-github — COMMENTED** [review]
Thanks for making the change and it's very helpful for our use case (https://github.com/facebook/rocksdb/issues/11407)!

File: db/memtable.cc
Comment (vrdhn) [reply db/memtable.cc:793]: My particular use case is read+range-delete, so this suffices for me.
I found that a write + range-delete workload is adversely affected by moving this check out of Read path

Can we find a middle ground ? 

Let me discuss with team to understand if we moving this out adversely affects our use case.

File: db/memtable.cc
Comment (vrdhn) [reply db/memtable.cc:174]: Okay, let me see if i can identify where to generate the flush reason.

File: include/rocksdb/options.h
Comment (vrdhn) [reply include/rocksdb/options.h:337]: Yes, that makes sense, i'll do that change.

File: options/cf_options.h
Comment (vrdhn) [reply options/cf_options.h:318]: I didn't updated the options/ , as we use Java api.
I'll update options.

File: db/memtable.cc
Comment (vrdhn) [reply db/memtable.cc:766]: The flag will be set in range-deletions, as requested.
Looks like our use case  will work well with it.

**Comment by vrdhn** [comment]
hi @yao-xiao-github ,@cbi42  I did the requested changes. I couldn't figure out where to set the FlushReason.
Any hints will be hightly appreciated.
Thanks.

**Comment by yao-xiao-github** [comment]
> hi @yao-xiao-github ,@cbi42 I did the requested changes. I couldn't figure out where to set the FlushReason.
> Any hints will be hightly appreciated.
> Thanks.

Thanks for making the changes. I also tried to expose the FlushReason but didn't find a good way. All the memtable related flushes seem to be put under the same reason. @cbi42 Could you provide some hints? Or is it possible to put info log here?

**Comment by cbi42** [comment]
> > hi @yao-xiao-github ,@cbi42 I did the requested changes. I couldn't figure out where to set the FlushReason.
> > Any hints will be hightly appreciated.
> > Thanks.
> 
> Thanks for making the changes. I also tried to expose the FlushReason but didn't find a good way. All the memtable related flushes seem to be put under the same reason. @cbi42 Could you provide some hints? Or is it possible to put info log here?

Hi - I also don't see an easy way to add FlushReason for this change, especially because ShouldFlushNow() was exclusively used to check if a memtable was full. Maybe we can just log the number of range tombstones here for now: https://github.com/facebook/rocksdb/blob/a5909f88641a1222865839e62c91e43e6ee36c03/db/flush_job.cc#L883

Also could you add a unit test to check that flush is triggered correctly based on the new option?

File: include/rocksdb/options.h
Comment (saintstack) [inline include/rocksdb/options.h:335]: Suggestion: If making another edit, you might say more than 'helps'; it is a guard against accumulating too many range deletes in memtable such that you get into a state where you cannot flush w/o OOM'ing.

File: options/options.cc
Comment (saintstack) [inline options/options.cc:453]: Good

**Review by saintstack — COMMENTED** [review]
Don't have perms to +1 this PR but LGTM.

Patched a 8.1.1 checkout then ran against a dataset that was stuck in recovery, OOM'ing every time because too many range deletes (Yao's case).  Ran patched binary with all defaults and still OOM. Then I set memtable_max_range_deletions=1000, recompiled and retried. Lots of flushes but was able to open the database.

**Comment by yao-xiao-github** [comment]
@cbi42 Could you please take another look? Thanks!

**Comment by cbi42** [comment]
> @cbi42 Could you please take another look? Thanks!

Hi - there are some requested changes like testing and logging that are not done. I have made these changes locally but realize this PR is not open for maintainer to edit yet. @vrdhn - can you update this PR to allow maintainer to edit? Thanks.

**Comment by bremac** [comment]
> > @cbi42 Could you please take another look? Thanks!
> 
> Hi - there are some requested changes like testing and logging that are not done. I have made these changes locally but realize this PR is not open for maintainer to edit yet. @vrdhn - can you update this PR to allow maintainer to edit? Thanks.

Hi @cbi42. Apparently [GitHub forbids opening PRs for a maintainer to edit on repos associated with an organization](https://github.com/orgs/community/discussions/5634). At @vrdhn's request I did make you a contributor on the repo. Let me know if that works for you. Otherwise, we can create a new PR from a repo associated with a personal account instead.

File: db/memtable.h
Comment (ajkr) [inline db/memtable.h:343]: This feels like a data race. What about comparing num_range_deletes_ against memtable_max_range_deletions_ in ShouldFlushNow()?

Comment (ajkr) [reply db/memtable.h:343]: Related (similar scenario where the flag started with one value and could only ever flip once to the other value): #4801

Comment (cbi42) [reply db/memtable.h]: Updated to check `num_range_deletes_` against `memtable_max_range_deletions_` in ShouldFlushNow().

**Review by ajkr — APPROVED** [review]
Looks great!

**Pull request closed without merging**
